### PR TITLE
DnsHelper for email check #10

### PR DESCRIPTION
**Pull request by kamarton**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | #10

### CODE REVIEW

File: src/DnsHelper.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    public static function checkMx(string $hostname): bool
+    {
+        $hostname .= '.';
+        try {
+            if (!@dns_check_record($hostname, 'MX')) {
+                return false;
+            }
+            $result = @dns_get_record($hostname, DNS_MX);
+            return $result !== false && count($result) > 0;
+        } catch (\Throwable $t) {
```
Comment (samdark) [inline src/DnsHelper.php]: If the intent is to ignore any exception then a comment indicating it should be added inside catch block.

File: src/DnsHelper.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
+<?php
+
+namespace Yiisoft\NetworkUtilities;
+
+class DnsHelper
+{
+    /**
+     * @param string $hostname hostname without dot at end
+     */
+    public static function checkMx(string $hostname): bool
```
Comment (samdark) [inline src/DnsHelper.php]:
```suggestion
    public static function checkMxExists(string $hostname): bool
```

File: src/DnsHelper.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+                return false;
+            }
+            $result = @dns_get_record($hostname, DNS_MX);
+            return $result !== false && count($result) > 0;
+        } catch (\Throwable $t) {
+        }
+        return false;
+    }
+
+    public static function checkA(string $hostname): bool
```
Comment (samdark) [inline src/DnsHelper.php]:
```suggestion
    public static function checkAExists(string $hostname): bool
```

File: src/DnsHelper.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+     * @param string $hostname hostname without dot at end
+     */
+    public static function checkMx(string $hostname): bool
+    {
+        $hostname .= '.';
+        try {
+            if (!@dns_check_record($hostname, 'MX')) {
+                return false;
+            }
+            $result = @dns_get_record($hostname, DNS_MX);
```
Comment (samdark) [inline src/DnsHelper.php]: Won't `dns_check_record` already check for existence? Why do we need `dns_get_record`?

Comment (samdark) [reply src/DnsHelper.php]: Same question about cases above.

File: src/DnsHelper.php (hunk 1 above)
Comment (samdark) [reply src/DnsHelper.php]: Same about other empty catch blocks.

File: src/DnsHelper.php [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+                return false;
+            }
+            $result = @dns_get_record($hostname, DNS_A);
+            return $result !== false && count($result) > 0;
+        } catch (\Throwable $t) {
+        }
+        return false;
+    }
+
+    public static function checkForEmail(string $hostnameOrEmail): bool
```
Comment (samdark) [inline src/DnsHelper.php]: I'd like to choose cleaner name like `isAcceptingServerForEmail`. @cebe, @SilverFire any suggestions?

File: src/DnsHelper.php (hunk 1 above)
Comment (kamarton) [reply src/DnsHelper.php]: corrected

File: src/DnsHelper.php (hunk 4 above)
Comment (kamarton) [reply src/DnsHelper.php]: corrected

File: src/DnsHelper.php [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+class DnsHelper
+{
+    /**
+     * @param string $hostname hostname without dot at end
+     * @link https://bugs.php.net/bug.php?id=78008
+     */
+    public static function checkMx(string $hostname): bool
+    {
+        $hostname .= '.';
```
Comment (SilverFire) [inline src/DnsHelper.php]: I'd make it fail-safe

```suggestion
        $hostname = rtrim($hostname, '.') . '.';
```

File: src/DnsHelper.php (hunk 5 above)
Comment (SilverFire) [reply src/DnsHelper.php]: `DnsHelper::isAcceptingServerForEmail()` makes me think about accepting of the DNS server. I'd not use the `server` word in this context.

Maybe `DnsHelper::domainAcceptsEmails()`?

Comment (kamarton) [reply src/DnsHelper.php]: > Maybe DnsHelper::domainAcceptsEmails()?

Like. Let me change it a bit: `DnsHelper::acceptsEmails($hostname)` or `DnsHelper::isAccpetsEmails()`. The whole helper is all about DNS, so I think the `domain` prefix is ​​just an overdetermination of the function.

File: src/DnsHelper.php (hunk 3 above)
Comment (samdark) [reply src/DnsHelper.php]: How about A and C records then? I think naming of these IP would add to confusion.

File: src/DnsHelper.php (hunk 5 above)
Comment (samdark) [reply src/DnsHelper.php]: `acceptsEmails()` sounds great

File: src/DnsHelper.php [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        } catch (\Throwable $t) {
+            // eg. name servers are not found https://github.com/yiisoft/yii2/issues/17602
+        }
+        return false;
+    }
+
+    /**
+     * @link https://tools.ietf.org/html/rfc5321#section-5
+     */
+    public static function domainAcceptsEmails(string $hostnameOrEmail): bool
```
Comment (samdark) [inline src/DnsHelper.php]:
```suggestion
    public static function acceptsEmails(string $hostnameOrEmail): bool
```

**Comment by samdark** [comment]
Merged. Thank you!

**Pull request merged**